src/net_processing.cpp

```diff
@@ -186,6 +186,53 @@ struct QueuedBlock {
     std::unique_ptr<PartiallyDownloadedBlock> partialBlock;
 };
 
+class TimeOffsets
+{
+private:
+    // Historic maximum number of time offsets we include in our median.
+    static constexpr size_t N{199};
+
+    mutable Mutex m_mutex;
+    std::array<int64_t, N> m_offsets GUARDED_BY(m_mutex){};
+    size_t m_index GUARDED_BY(m_mutex){0};
+
+public:
+    /** Add a new time offset sample. */
+    void Add(int64_t offset) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+
+        // Stop sampling after N samples.
+        // TODO This maintains a historic bug of adjusted time and could be
+        // changed in the future.
+        if (m_index >= N) return;
+
+        m_offsets[m_index] = offset;
+        ++m_index;
+    }
+
+    /** Compute and return the median of the collected time offset samples. */
+    int64_t Median() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+
+        // Only compute the median from 5 or more samples.
+        if (m_index < 4) return 0;
+
+        std::array<int64_t, N> sorted_copy = m_offsets;
+        std::sort(sorted_copy.begin(), sorted_copy.end());
+
+        if (m_index % 2 == 0) {
+            return sorted_copy[m_index / 2];
+        } else {
+            return (sorted_copy[m_index / 2] / 2) +
+                   (sorted_copy[m_index / 2 - 1] / 2);
+        }
+    }
```

Comment:
> I think the calculation is wrong: the average of the 2 middle elements should be taken in case of an even number of elements.

yes🤦

> UB: sorted_copy being an array of size N, I think we'd be sorting uninitialized values if m_index < N?

No, `m_offsets` is not uninitialized

Reply:
> No, `m_offsets` is not uninitialized

The array is initialized, but the items are not, and afaik `int64_t` doesn't have a default constructor?

Reply:
https://godbolt.org/z/5PKET7E39

Reply:
Ah cool, didn't know, thanks. Still, having default (0-)values in sorted_copy seems like it would lead to an incorrect median calculation?